lib/pbench/server/auth/__init__.py

```diff
@@ -0,0 +1,441 @@
+        Returns:
+            Extracted token information
+            {
+                "aud": <targeted audience id>,
+                "email_verified": <true-or-false>,
+                "expires_in": <Number_of_seconds>,
+                "access_type": "offline",
+                "exp": <unix_timestamp>,
+                "azp": <client_id>,
+                "scope": "openid email profile",
+                "email": <user_email>,
+                "sub": <user_id>
+            }
+        """
+        payload = {
+            "client_id": self.client_id,
+            "client_secret": self.client_secret_key,
+            "token": token,
+        }
+        return self._post(token_info_uri, data=payload).json()
+
+    def token_introspect_offline(
+        self,
+        token: str,
+        key: str,
+        audience: str = "account",
+        algorithms: List[str] = ["RS256"],
+        **kwargs,
+    ) -> JSON:
+        """
+        Utility method to decode access/Id tokens using the public key provided
+        by the identity provider.
+        The introspected JWT token contains the claims specified in
+        https://tools.ietf.org/html/rfc7662
+
+        Please refer https://www.rfc-editor.org/rfc/rfc7662.html#section-4 for
+        reasons on doing the token introspection offline.
+        Args:
+            token: token value to introspect
+            key: client public key
+            audience: jwt token audience/client, who this token was intended for
+            algorithms: Algorithm with which this JWT token was encoded
+        Returns:
+            Extracted token information
+            {
+                "aud": <targeted audience id>, # client_id
+                "email_verified": <true-or-false>,
+                "expires_in": <Number_of_seconds>,
+                "access_type": "offline",
+                "exp": <unix_timestamp>,
+                "azp": <client_id>,
+                "scope": "openid email profile",
+                "email": <user_email>,
+                "sub": <user_id>
+            }
+        """
+        return jwt.decode(
+            token, key, algorithms=algorithms, audience=audience, **kwargs
+        )
+
+    def get_userinfo(self, token: str = None) -> JSON:
+        """
+        The userinfo endpoint returns standard claims about the authenticated
+        user, and is protected by a bearer token.
+        http://openid.net/specs/openid-connect-core-1_0.html#UserInfo
+        Args:
+            token: Valid token to extract the userinfo
+        Returns:
+            Userinfo payload
+            {
+                "family_name": <surname>,
+                "sub": <user_id>,
+                "picture": <URL>,
+                "locale": <locale-name>,
+                "email_verified": <true-or-false>,
+                "given_name": <given-name>,
+                "email": <email-address>,
+                "hd": <domain name>,
+                "name": <full-name>
```

Comment:
My eyes skipped over this several times, I think, but I debarked on a mission to track down all my open conversations to figure out what I can close.

I like that you replaced all the sample values with `<...>` tags containing some descriptive text, so I can resolve that conversation...

However, there's a lot of inconsistency around the notation here. We have `<blank separated words>`, `<hyphen-separated-words>` and `<underscore_separated_words>` and it'd be nice if they were all consistent so readers don't have to think "wait, does this difference *matter*?"

I happen to prefer the hyphen-separated-style myself, but that's less important than being consistent. 👀

Reply:
That's a good point, I should make them consistent

Reply:
Okay, I decided to go with an underscore, since the number of changes I had to make was much less.